File: Documentation/4.0/Samples/java/src/test/java/net/ravendb/ClientApi/Session/SavingChanges.java
package net.ravendb.ClientApi.Session;

import net.ravendb.client.documents.DocumentStore;
import net.ravendb.client.documents.IDocumentStore;
import net.ravendb.client.documents.session.IDocumentSession;

import java.time.Duration;

public class SavingChanges {

    private interface IFoo {
        //region saving_changes_1
        void saveChanges();
        //endregion
    }

    private class Employee {
        private String firstName;
        private String lastName;

        public String getFirstName() {
            return firstName;
        }

        public void setFirstName(String firstName) {
            this.firstName = firstName;
        }

        public String getLastName() {
            return lastName;
        }

        public void setLastName(String lastName) {
            this.lastName = lastName;
        }
    }

    public SavingChanges() {
        try (IDocumentStore store = new DocumentStore()) {
            try (IDocumentSession session = store.openSession()) {
                //region saving_changes_2
                Employee employee = new Employee();
                employee.setFirstName("John");
                employee.setLastName("Doe");

                session.store(employee);
                session.saveChanges();;
                //endregion
            }

            try (IDocumentSession session = store.openSession()) {
                //region saving_changes_3
                session.advanced().waitForIndexesAfterSaveChanges(builder -> {
                    builder.withTimeout(Duration.ofSeconds(30))
                        .throwOnTimeout(true)
                        .waitForIndexes("index/1", "index/2");

                    Employee employee = new Employee();
                    employee.setFirstName("John");
                    employee.setLastName("Doe");
                    session.store(employee);

                    session.saveChanges();
                });
                //endregion
            }

            try (IDocumentSession session = store.openSession()) {
                //region saving_changes_4
                session
                    .advanced()
                    .waitForReplicationAfterSaveChanges(builder -> {
                        builder.withTimeout(Duration.ofSeconds(30))
                            .throwOnTimeout(false) //default true
                            .numberOfReplicas(2)//minimum replicas to replicate
                            .majority(false);
                    });

                Employee employee = new Employee();
                employee.setFirstName("John");
                employee.setLastName("Doe");

                session.store(employee);
                session.saveChanges();
                //endregion
            }
        }
    }
}

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/SavingChanges.cs
using System;
using System.Threading.Tasks;
using Raven.Client.Documents;
using Raven.Client.Documents.Session;
using Raven.Documentation.Samples.Orders;
using Xunit.Sdk;

namespace Raven.Documentation.Samples.ClientApi.Session
{
    public class SavingChanges
    {
        private interface IInterface
        {
            #region saving_changes_1
            void SaveChanges();
            #endregion

            #region saving_changes_1_async
            Task SaveChangesAsync();
            #endregion
        }

        public async Task SavingChangesXY()
        {
            using (var store = new DocumentStore())
            {
                using (var session = store.OpenSession())
                {
                    #region saving_changes_2
                    // storing new entity
                    session.Store(new Employee
                    {
                        FirstName = "John",
                        LastName = "Doe"
                    });

                    session.SaveChanges();
                    #endregion
                }
            }

            using (var store = new DocumentStore())
            {
                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region saving_changes_2_async
                    // storing new entity
                    await asyncSession.StoreAsync(new Employee
                    {
                        FirstName = "John",
                        LastName = "Doe"
                    });

                    await asyncSession.SaveChangesAsync();
                    #endregion
                }
            }


            using (var store = new DocumentStore())
            {
                using (var session = store.OpenSession())
                {
                    // storing new entity
                    #region saving_changes_3
                    session.Advanced.WaitForIndexesAfterSaveChanges(
                        timeout: TimeSpan.FromSeconds(30),
                        throwOnTimeout: true,
                        indexes: new[] { "index/1", "index/2" });

                    session.Store(new Employee
                    {
                        FirstName = "John",
                        LastName = "Doe"
                    });

                    session.SaveChanges();
                    #endregion
                }
            }

            using (var store = new DocumentStore())
            {
                using (var asyncSession = store.OpenAsyncSession())
                {
                    // storing new entity
                    #region saving_changes_3_async
                    asyncSession.Advanced.WaitForIndexesAfterSaveChanges(
                        timeout: TimeSpan.FromSeconds(30),
                        throwOnTimeout: true,
                        indexes: new[] { "index/1", "index/2" });

                    await asyncSession.StoreAsync(new Employee
                    {
                        FirstName = "John",
                        LastName = "Doe"
                    });

                    await asyncSession.SaveChangesAsync();
                    #endregion
                }
            }

            using (var store = new DocumentStore())
            {
                using (var session = store.OpenSession())
                {
                    // storing new entity
                    #region saving_changes_4

                    session.Advanced.WaitForReplicationAfterSaveChanges(
                        timeout: TimeSpan.FromSeconds(30),
                        throwOnTimeout: false, //default true
                        replicas: 2, //minimum replicas to replicate
                        majority: false);

                    session.Store(new Employee
                    {
                        FirstName = "John",
                        LastName = "Doe"
                    });

                    session.SaveChanges();
                    #endregion
                }
            }

            using (var store = new DocumentStore())
            {
                using (var asyncSession = store.OpenAsyncSession())
                {
                    // storing new entity
                    #region saving_changes_4_async

                    asyncSession.Advanced.WaitForReplicationAfterSaveChanges(
                        timeout: TimeSpan.FromSeconds(30),
                        throwOnTimeout: false, //default true
                        replicas: 2, //minimum replicas to replicate
                        majority: false);

                    await asyncSession.StoreAsync(new Employee
                    {
                        FirstName = "John",
                        LastName = "Doe"
                    });

                    await asyncSession.SaveChangesAsync();
                    #endregion
                }
            }

            #region cluster_store_with_compare_exchange
            using (var store = new DocumentStore())
            {
                using (var session = store.OpenSession(new SessionOptions
                {
                    //default is:     TransactionMode.SingleNode
                    TransactionMode = TransactionMode.ClusterWide
                }))
                {
                    var user = new Employee
                    {
                        FirstName = "John",
                        LastName = "Doe"
                    };
                    session.Store(user);

                    // this transaction is now conditional on this being 
                    // successfully created (so, no other users with this name)
                    // it also creates an association to the new user's id
                    session.Advanced.ClusterTransaction
                        .CreateCompareExchangeValue("usernames/John", user.Id);

                    session.SaveChanges();
                }
                #endregion

                #region cluster_store_with_compare_exchange_async
                using (var session = store.OpenAsyncSession(new SessionOptions
                {
                    //default is:     TransactionMode.SingleNode
                    TransactionMode = TransactionMode.ClusterWide
                }))
                {
                    var user = new Employee
                    {
                        FirstName = "John",
                        LastName = "Doe"
                    };
                    await session.StoreAsync(user);

                    // this transaction is now conditional on this being 
                    // successfully created (so, no other users with this name)
                    // it also creates an association to the new user's id
                    session.Advanced.ClusterTransaction
                        .CreateCompareExchangeValue("usernames/John", user.Id);

                    await session.SaveChangesAsync();
                }
                #endregion
            }
        }
    }
}
